### ISSUE TITLE
DataRobot integration

### ISSUE THREAD

**Issue by pavloko** [looker/actions#183]
Hi!

I’m Pavlo an engineer from DataRobot working on integration with Looker.

I’d like to double check our current implementation as it sends data in a way that was not used in other actions.

We had difficulties sending `readable` stream as part `multipart/form-data` request. The problem lies in `form-data` library that `request-promise-native` uses under the hood, and it cannot process `passThrough` streams at the moment. Here’s related [issue](https://github.com/request/request/issues/2499).

As a workaround, we’re sending `request.scheduledPlan.downloadUrl` to DataRobot servers. The server then fetches the data directly from this url as it appears to be publicly available.

Do you think there could be potential security issues with this approach?

  ```ts
async execute(request: Hub.ActionRequest) {
    const url = request.scheduledPlan && request.scheduledPlan.downloadUrl
    const options = {
      url: "https://app.datarobot.com/api/v2/projects/",
      headers: {
        Authorization: `Token ${request.params.apiToken}`,
      },
    }
    if (url) {
      Object.assign(options, {
        body: {
          projectName: request.formParams.projectName,
          url: request.scheduledPlan && request.scheduledPlan.downloadUrl,
        },
        json: true,
      })
    } else {
      Object.assign(options, {
        formData: {
          file: {
            value: request.attachment && request.attachment.dataBuffer,
            options: {
              filename: request.suggestedFilename(),
              contentType: "text/csv",
            },
          },
          projectName: request.formParams.projectName,
        },
      })
    }

    try {
      await httpRequest.post(options).promise()
      return new Hub.ActionResponse({ success: true })
    } catch (e) {
      return new Hub.ActionResponse({ success: false, message: e.message })
    }
  }
```

**Comment by wilg** [comment]
I am not certain if it's a security issue – it's a time sensitive secret URL so it might be okay.

The main issue I see with this approach is that Looker does not always send a URL – sometimes it sends the data inline, so you'd need to implement both approaches which would be very brittle. The DataActionRequest tries to shield you from knowing whether Looker sent a URL or not.

I'd try using an alternate HTTP library instead of `request`, like `bhttp` mentioned in the issue you referenced. There's no real requirement to use the same HTTP library that's used in the other actions.

**Comment by pavloko** [comment]
Hey @wilg. Thanks you for the reply and apologize for the delay with the response. 

I understand your concern about this, and really like the way `request.stream` works. 
Unfortunately, I wasn't able to make the `formData` work together with the `PassThrough` stream. I've tried the following libraries:
1. https://github.com/octet-stream/form-data
2. https://github.com/joepie91/node-form-data2
3. https://www.npmjs.com/package/form-data. Fork that supposedly should allow custom streams.

I've submitted the [PR](https://github.com/looker/actions/pull/187) with the full code of our action and would like to hear your comments.

Thanks.
